File: divide_conquer/histogram.py
# https://www.acmicpc.net/problem/6549

array = []

while True:
    arr = input()
    if arr == '0':
        break
    array.append(list(map(int, arr.split())))

# def histogram(array):
#     if len(array) == 1:
#         return array[0]

#     value = len(array) * min(array)
#     left, right = histogram(array[:-1]), histogram(array[1:])

#     return max([value, left, right])

def histogram(array):
    answer = []
    maxHeight = 0

    for idx, val in enumerate(array):
        if len(answer) == 0:
            answer.append([idx, val])
            continue
        flag = False
        temp = None
        while len(answer) > 0 and answer[-1][1] > val:
            flag = True
            last = answer.pop()
            maxHeight = max(maxHeight, last[1] * (idx - last[0]))
            temp = [last[0], val]
        if flag:
            answer.append(temp)
        answer.append([idx, val])
    
    for item in answer:
        maxHeight = max(maxHeight, item[1] * (len(array) - item[0]))
    return maxHeight

for arr in array:
    print(histogram(arr[1:]))

File: data_structure/histogram.js
// https://www.acmicpc.net/problem/1725

const stdin = require("fs")
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : __dirname + "/../input.txt")
  .toString()
  .trim()
  .split("\n");
const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

const N = +input();
const heights = Array.from({ length: N }, () => +input());
let answer = 0;
let stack = [];

for (let index in heights) {
  const height = heights[index];
  let temp;
  while (stack.length && stack.at(-1)[1] > height) {
    const [curIndex, curHeight] = stack.pop();
    const area = (index - curIndex) * curHeight;
    answer = Math.max(area, answer);
    temp = [curIndex, height];
  }
  if (temp) stack.push(temp);
  stack.push([index, height]);
}

while (stack.length) {
  const [curIndex, curHeight] = stack.pop();
  const area = (N - curIndex) * curHeight;
  answer = Math.max(area, answer);
}

console.log(answer);
